crashtracker/src/collector/api.rs

```diff
@@ -593,3 +593,158 @@ fn test_altstack_nouse() -> anyhow::Result<()> {
     // OK, we're done
     Ok(())
 }
+
+#[cfg_attr(miri, ignore)]
+#[cfg(target_os = "linux")]
+#[test]
+fn test_waitall_nohang() -> anyhow::Result<()> {
+    // This test checks whether the crashtracking implementation can cause malformed `waitall()`
+    // idioms to hang.
+    // Consider the following code from the Ruby runtime:
+    //
+    //   static VALUE
+    //   proc_waitall(VALUE _)
+    //   {
+    //       VALUE result;
+    //       rb_pid_t pid;
+    //       int status;
+    //
+    //       result = rb_ary_new();
+    //       rb_last_status_clear();
+    //
+    //       for (pid = -1;;) {
+    //           pid = rb_waitpid(-1, &status, 0);
+    //           if (pid == -1) {
+    //               int e = errno;
+    //               if (e == ECHILD)
+    //                   break;
+    //               rb_syserr_fail(e, 0);
+    //           }
+    //           rb_ary_push(result, rb_assoc_new(PIDT2NUM(pid), rb_last_status_get()));
+    //       }
+    //       return result;
+    //   }
+    //
+    // The intent here is to wait for all of one's child processes to exit.  This is a pretty
+    // standard operation in multi-process situations, with one important caveat:  usually you know
+    // your children ahead of time and can wait on them in a controlled, intentional matter.
+    // Previous versions of crashtracking, which spawned long-lived receiver processes, would
+    // interfere with this
+    //
+    // This implements the inner behavior of a test which allows the caller to control which
+    // options are used.
+    use crate::StacktraceCollection;
+    use chrono::Utc;
+    use ddcommon::Endpoint;
+
+    let time = Utc::now().to_rfc3339();
+    let dir = "/tmp/crashreports/";
+    let output_url = format!("file://{dir}{time}.txt");
+
+    let endpoint = Some(Endpoint::from_slice(&output_url));
+
+    let path_to_receiver_binary =
+        "/tmp/libdatadog/bin/libdatadog-crashtracking-receiver".to_string();
+    let create_alt_stack = false;
+    let use_alt_stack = false;
```

Comment:
I can't think of a reason why it would matter, and when I originally wrote the test I thought that was sufficient reason to avoid using the altstack ("don't test irregular configs"), but in practice almost all of our configurations run on the altstack, so we should test that as the default.

Changing--good callout